File: implementation/scipamato/core/sync/src/main/java/ch/difty/scipamato/core/sync/jobs/newsletter/NewStudyItemWriter.java
package ch.difty.scipamato.core.sync.jobs.newsletter;

import static ch.difty.scipamato.publ.db.tables.NewStudy.NEW_STUDY;

import org.jooq.DSLContext;
import org.springframework.batch.item.ItemWriter;

import ch.difty.scipamato.core.sync.jobs.ScipamatoItemWriter;

/**
 * {@link ItemWriter} implementation to synchronize newStudies from scipamato-core
 * to public.
 *
 * <ul>
 * <li>Takes care of inserts and updates.</li>
 * <li>Every record will be updated at least with the current timestamp in
 * last_synched.</li>
 * <li>id columns and audit fields are those of the scipamato-core tables</li>
 * </ul>
 *
 * @author u.joss
 */
public class NewStudyItemWriter extends ScipamatoItemWriter<PublicNewStudy> {

    public NewStudyItemWriter(final DSLContext jooqDslContextPublic) {
        super(jooqDslContextPublic, "newStudy");
    }

    @Override
    protected int executeUpdate(final PublicNewStudy p) {
        return getDslContext()
            .insertInto(NEW_STUDY)
            .columns(NEW_STUDY.NEWSLETTER_ID, NEW_STUDY.NEWSLETTER_TOPIC_ID, NEW_STUDY.SORT, NEW_STUDY.PAPER_NUMBER,
                NEW_STUDY.YEAR, NEW_STUDY.AUTHORS, NEW_STUDY.HEADLINE, NEW_STUDY.DESCRIPTION, NEW_STUDY.VERSION,
                NEW_STUDY.CREATED, NEW_STUDY.LAST_MODIFIED, NEW_STUDY.LAST_SYNCHED)
            .values(p.getNewsletterId(), p.getNewsletterTopicId(), p.getSort(), p.getPaperNumber(), p.getYear(),
                p.getAuthors(), p.getHeadline(), p.getDescription(), p.getVersion(), p.getCreated(),
                p.getLastModified(), p.getLastSynched())
            .onConflict(NEW_STUDY.NEWSLETTER_ID, NEW_STUDY.PAPER_NUMBER)
            .doUpdate()
            .set(NEW_STUDY.NEWSLETTER_TOPIC_ID, p.getNewsletterTopicId())
            .set(NEW_STUDY.SORT, p.getSort())
            .set(NEW_STUDY.PAPER_NUMBER, p.getPaperNumber())
            .set(NEW_STUDY.YEAR, p.getYear())
            .set(NEW_STUDY.AUTHORS, p.getAuthors())
            .set(NEW_STUDY.HEADLINE, p.getHeadline())
            .set(NEW_STUDY.DESCRIPTION, p.getDescription())
            .set(NEW_STUDY.VERSION, p.getVersion())
            .set(NEW_STUDY.CREATED, p.getCreated())
            .set(NEW_STUDY.LAST_MODIFIED, p.getLastModified())
            .set(NEW_STUDY.LAST_SYNCHED, p.getLastSynched())
            .where(NEW_STUDY.NEWSLETTER_ID
                .eq(p.getNewsletterId())
                .and(NEW_STUDY.PAPER_NUMBER.eq(p.getPaperNumber())))
            .execute();
    }

}

File: core/core-sync/src/main/kotlin/ch/difty/scipamato/core/sync/jobs/newsletter/NewStudyItemWriter.kt
package ch.difty.scipamato.core.sync.jobs.newsletter

import ch.difty.scipamato.core.sync.PublicNewStudy
import ch.difty.scipamato.core.sync.jobs.ScipamatoItemWriter
import ch.difty.scipamato.publ.db.tables.NewStudy
import org.jooq.DSLContext

/**
 * [ScipamatoItemWriter] implementation to synchronize newStudies from scipamato-core to public.
 *
 *  * Takes care of inserts and updates.
 *  * Every record will be updated at least with the current timestamp in last_synched.
 *  * id columns and audit fields are those of the scipamato-core tables
 */
class NewStudyItemWriter(jooqDslContextPublic: DSLContext) : ScipamatoItemWriter<PublicNewStudy>(jooqDslContextPublic, "newStudy") {
    override fun executeUpdate(i: PublicNewStudy): Int =
        dslContext
            .insertInto(NewStudy.NEW_STUDY)
            .columns(NewStudy.NEW_STUDY.NEWSLETTER_ID, NewStudy.NEW_STUDY.NEWSLETTER_TOPIC_ID, NewStudy.NEW_STUDY.SORT,
                NewStudy.NEW_STUDY.PAPER_NUMBER, NewStudy.NEW_STUDY.YEAR, NewStudy.NEW_STUDY.AUTHORS, NewStudy.NEW_STUDY.HEADLINE,
                NewStudy.NEW_STUDY.DESCRIPTION, NewStudy.NEW_STUDY.VERSION, NewStudy.NEW_STUDY.CREATED, NewStudy.NEW_STUDY.LAST_MODIFIED,
                NewStudy.NEW_STUDY.LAST_SYNCHED)
            .values(i.newsletterId, i.newsletterTopicId, i.sort, i.paperNumber, i.year,
                i.authors, i.headline, i.description, i.version, i.created,
                i.lastModified, i.lastSynched)
            .onConflict(NewStudy.NEW_STUDY.NEWSLETTER_ID, NewStudy.NEW_STUDY.PAPER_NUMBER)
            .doUpdate()
            .set(NewStudy.NEW_STUDY.NEWSLETTER_TOPIC_ID, i.newsletterTopicId)
            .set(NewStudy.NEW_STUDY.SORT, i.sort)
            .set(NewStudy.NEW_STUDY.PAPER_NUMBER, i.paperNumber)
            .set(NewStudy.NEW_STUDY.YEAR, i.year)
            .set(NewStudy.NEW_STUDY.AUTHORS, i.authors)
            .set(NewStudy.NEW_STUDY.HEADLINE, i.headline)
            .set(NewStudy.NEW_STUDY.DESCRIPTION, i.description)
            .set(NewStudy.NEW_STUDY.VERSION, i.version)
            .set(NewStudy.NEW_STUDY.CREATED, i.created)
            .set(NewStudy.NEW_STUDY.LAST_MODIFIED, i.lastModified)
            .set(NewStudy.NEW_STUDY.LAST_SYNCHED, i.lastSynched)
            .where(NewStudy.NEW_STUDY.NEWSLETTER_ID.eq(i.newsletterId).and(NewStudy.NEW_STUDY.PAPER_NUMBER.eq(i.paperNumber)))
            .execute()
}
